`no-empty-description` incorrectly flags non-empty descriptions.

The following seem like reasonable test descriptions, but they get flagged by `no-empty-description`:

```js
describe('eslint-plugin-mocha', function() {
  const description = 'description';
  it(description, function() {});

  ['test1', 'test2'].forEach(testname => {
    it(testname, function() {});
  });

  const removePrefix = 'testcase.prefix';
  it(removePrefix.replace('.prefix'), function() {});
});
```

```
   3:3  error  Unexpected empty test description  mocha/no-empty-description
   6:5  error  Unexpected empty test description  mocha/no-empty-description
  10:3  error  Unexpected empty test description  mocha/no-empty-description
```

Was this intended with #304?

--- a/lib/rules/no-empty-description.js
+++ b/lib/rules/no-empty-description.js
@@ -22,7 +22,7 @@ function isIdentifier(node) {
     return node && node.type === 'Identifier';
 }
 
-function isStaticallyAnalyzableTitle(node, extractedText) {
+function isStaticallyAnalyzableDescription(node, extractedText) {
     if (extractedText === null) {
         return !(isTemplateString(node) || isIdentifier(node));
     }
@@ -71,7 +71,7 @@ module.exports = {
             const description = mochaCallExpression.arguments[0];
             const text = getStringIfConstant(description, context.getScope());
 
-            if (!isStaticallyAnalyzableTitle(description, text)) {
+            if (!isStaticallyAnalyzableDescription(description, text)) {
                 return true;
             }
 

--- a/lib/rules/no-empty-description.js
+++ b/lib/rules/no-empty-description.js
@@ -67,7 +67,7 @@ module.exports = {
             return node.callee && node.callee.name && testNames.includes(node.callee.name);
         }
 
-        function checkDescription(mochaCallExpression) {
+        function isNonEmptyDescription(mochaCallExpression) {
             const description = mochaCallExpression.arguments[0];
             const text = getStringIfConstant(description, context.getScope());
 
@@ -81,7 +81,7 @@ module.exports = {
         return {
             CallExpression(node) {
                 if (isTest(node)) {
-                    if (!checkDescription(node)) {
+                    if (!isNonEmptyDescription(node)) {
                         context.report({
                             node,
                             message: message || ERROR_MESSAGE

--- a/lib/rules/no-empty-description.js
+++ b/lib/rules/no-empty-description.js
@@ -1,6 +1,6 @@
 'use strict';
 
-const { getStringIfConstant } = require('eslint-utils');
+const { getStringIfConstant, getStaticValue } = require('eslint-utils');
 
 const DEFAULT_TEST_NAMES = [ 'describe', 'context', 'suite', 'it', 'test', 'specify' ];
 const ERROR_MESSAGE = 'Unexpected empty test description.';
@@ -55,11 +55,19 @@ module.exports = {
             return [ 'TaggedTemplateExpression', 'TemplateLiteral' ].includes(node && node.type);
         }
 
+        function isIdentifier(node) {
+            return node && node.type === 'Identifier';
+        }
+
         function checkDescription(mochaCallExpression) {
             const description = mochaCallExpression.arguments[0];
-            const text = getStringIfConstant(description);
+            const text = getStringIfConstant(description, context.getScope());
 
             if (isTemplateString(description) && text === null) {
+                return true;
+            }
+
+            if (isIdentifier(description) && text === null) {
                 return true;
             }
 

--- a/lib/rules/no-empty-description.js
+++ b/lib/rules/no-empty-description.js
@@ -1,6 +1,6 @@
 'use strict';
 
-const { getStringIfConstant, getStaticValue } = require('eslint-utils');
+const { getStringIfConstant } = require('eslint-utils');
 
 const DEFAULT_TEST_NAMES = [ 'describe', 'context', 'suite', 'it', 'test', 'specify' ];
 const ERROR_MESSAGE = 'Unexpected empty test description.';
@@ -12,6 +12,14 @@ function objectOptions(options = {}) {
     } = options;
 
     return { testNames, message };
+}
+
+function isTemplateString(node) {
+    return [ 'TaggedTemplateExpression', 'TemplateLiteral' ].includes(node && node.type);
+}
+
+function isIdentifier(node) {
+    return node && node.type === 'Identifier';
 }
 
 module.exports = {
@@ -49,14 +57,6 @@ module.exports = {
 
         function isTest(node) {
             return node.callee && node.callee.name && testNames.includes(node.callee.name);
-        }
-
-        function isTemplateString(node) {
-            return [ 'TaggedTemplateExpression', 'TemplateLiteral' ].includes(node && node.type);
-        }
-
-        function isIdentifier(node) {
-            return node && node.type === 'Identifier';
         }
 
         function checkDescription(mochaCallExpression) {

--- a/lib/rules/no-empty-description.js
+++ b/lib/rules/no-empty-description.js
@@ -22,6 +22,14 @@ function isIdentifier(node) {
     return node && node.type === 'Identifier';
 }
 
+function isStaticallyAnalyzableTitle(node, extractedText) {
+    if (extractedText === null) {
+        return !(isTemplateString(node) || isIdentifier(node));
+    }
+
+    return true;
+}
+
 module.exports = {
     meta: {
         type: 'suggestion',
@@ -63,11 +71,7 @@ module.exports = {
             const description = mochaCallExpression.arguments[0];
             const text = getStringIfConstant(description, context.getScope());
 
-            if (isTemplateString(description) && text === null) {
-                return true;
-            }
-
-            if (isIdentifier(description) && text === null) {
+            if (!isStaticallyAnalyzableTitle(description, text)) {
                 return true;
             }
 

--- a/test/rules/no-empty-description.js
+++ b/test/rules/no-empty-description.js
@@ -29,6 +29,9 @@ ruleTester.run('no-empty-description', rules['no-empty-description'], {
         'test.only("some text")',
         'test("some text", function() { })',
 
+        'var dynamicTitle = "foo"; it(dynamicTitle, function() {});',
+        'it(dynamicTitle, function() {});',
+
         'notTest()',
 
         {
@@ -46,6 +49,10 @@ ruleTester.run('no-empty-description', rules['no-empty-description'], {
         {
             options: [ { testNames: [ 'someFunction' ] } ],
             code: 'someFunction("this is a test", function () { });'
+        },
+        {
+            parserOptions: { ecmaVersion: 2019 },
+            code: 'const dynamicTitle = "foo"; it(dynamicTitle, function() {});'
         }
     ],
 
@@ -76,6 +83,11 @@ ruleTester.run('no-empty-description', rules['no-empty-description'], {
             parserOptions: { ecmaVersion: 2019 },
             code: 'it(` `, function () { });',
             errors: [ { message: defaultErrorMessage, ...firstLine } ]
+        },
+        {
+            parserOptions: { ecmaVersion: 2019 },
+            code: 'const foo = ""; it(foo);',
+            errors: [ { message: defaultErrorMessage, line: 1, column: 17 } ]
         }
     ]
 

